sklearn/preprocessing/label.py

```diff
@@ -891,14 +891,27 @@ def transform(self, y):
         """
         check_is_fitted(self, 'classes_')
 
-        class_to_index = dict(zip(self.classes_, range(len(self.classes_))))
+        class_to_index = self._cached_dict_for()
         yt = self._transform(y, class_to_index)
 
         if not self.sparse_output:
             yt = yt.toarray()
 
         return yt
 
+    def _cached_dict_for(self):
+        if not hasattr(self, '_prev_classes'):
+            self._prev_classes = self.classes_
+            self._cached_dict = None
+
+        if (self._prev_classes.__repr__() != self.classes_.__repr__()) \
```

Comment:
This `__repr__` comparison would miss subtle changes in the classes, wouldn't it?
I would suggest to just reset `self._cached_dict` at each `fit` and `fit_transform`.

Reply:
you mean like that? 36dc1a4